### PR TITLE
Embed luarocks in static build

### PR DESCRIPTION
**Pull request by locker**
This PR embeds luarocks and lua-zlib, which is used by luarocks, into static builds. It's also possible to embed them into regular builds by setting `EMBED_LUAROCKS` and `EMBED_LUAZLIB` cmake options. This is needed to make statically built tarantool binary truly independent of the environment it runs in (currently, one has to install luarocks for `tarantoolctl rocks` to work). The patches were pulled from the [enterprise version](https://github.com/tarantool/enterprise/commit/b77f4cbc28659aeb4a2ef671f548bc19737925f3) with some minor changes:
- Introduction of `FindZLIB.cmake`
- `EMBED_LUAZLIB` and `EMBED_LUAROCKS` cmake options.
- gitignore of `lua.c` files for in-source builds.
- Refactoring of `lua_source`.
- Skip installation of luarocks if it's embedded.
- Drop luazip. Luarocks doesn't need it. Looks like it's another enterprise dependency.
- Split the patch for review.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -46,3 +46,6 @@
 [submodule "third_party/c-dt"]
 	path = third_party/c-dt
 	url = https://github.com/tarantool/c-dt.git
+[submodule "third_party/lua-zlib"]
+	path = third_party/lua-zlib
+	url = https://github.com/tarantool/lua-zlib.git
```
Comment (Totktonada) [inline .gitmodules:52]: We bundle it into our source tarball, so we should put the copyright holder and license into `debian/copyright`.

**Comment by Totktonada** [comment]
AFAIU, we need lua-zlib for `pack` and `unpack` subcommands: luarocks [implements](https://github.com/tarantool/luarocks/blob/954c2f5784cedd7d70f1b0f5d6af6e36c21e94be/src/luarocks/tools/zip.lua) ZIP files (de)compression of top of it. I guess it also may be used in the `install` subcommand, when rockspec points to a source tarball in [source.url](https://github.com/luarocks/luarocks/wiki/Rockspec-format#build-rules)[^1].

Okay. What I don't understand: why do we need luazip (libzzip backed implementation of ZIP file decompression)?

[^1]: Not directly relevant here, but for the full picture: if the zlib backed implementation cannot be used, luarocks [fallbacks](https://github.com/tarantool/luarocks/blob/954c2f5784cedd7d70f1b0f5d6af6e36c21e94be/src/luarocks/fs.lua#L81-L90) to `zip` / `unzip` commands [execution](https://github.com/tarantool/luarocks/blob/954c2f5784cedd7d70f1b0f5d6af6e36c21e94be/src/luarocks/fs/unix/tools.lua#L122-L153).

File: .gitmodules (hunk 1 above)
Comment (locker) [reply .gitmodules:52]: Fixed.

**Comment by locker** [comment]
> Okay. What I don't understand: why do we need luazip (libzzip backed implementation of ZIP file decompression)?

You're right - we don't need it for luarocks to work. Probably, it's yet another enterprise dependency that got mixed in. I removed the corresponding patch from this PR. I'll submit it separately if it turns out we still need it.

**Review by Totktonada — APPROVED** [review]
LGTM. I glanced over the patchset and aside of highlighted (and already fixed) comments nothing confuses me. I don't think that it would be profitable to dive deeper into details.

**Comment by olegrok** [comment]
> > Okay. What I don't understand: why do we need luazip (libzzip backed implementation of ZIP file decompression)?
> 
> You're right - we don't need it for luarocks to work. Probably, it's yet another enterprise dependency that got mixed in. I removed the corresponding patch from this PR. I'll submit it separately if it turns out we still need it.

I guess it's pure enterprise dependency. I think there were plans that cartridge will support config upload as zip archive.
That's a reason why such dependency was added. However AFAIK only TDG uses it.

**Pull request merged**